File: solution/979. Distribute Coins in Binary Tree/979.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int distributeCoins(TreeNode* root) {
        int res = 0;
        balance(root, res);
        return res;
    }
    
    int balance(TreeNode* node, int& res) {
        if (!node) return 0;
        int l = balance(node->left, res);
        int r = balance(node->right, res);
        res += abs(l) + abs(r);
        return l + r + node->val - 1;
    }
};

File: solution/979. Distribute Coins in Binary Tree/979.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    res = 0
    
    def distributeCoins(self, root: Optional[TreeNode]) -> int:
        def balance(node):
            if not node: return 0
            l = balance(node.left)
            r = balance(node.right)
            self.res += abs(l) + abs(r)
            return l + r + node.val - 1
        
        balance(root)
        return self.res
    
